### DISCUSSION TITLE
How to load only my own records (user_id filter)

### DISCUSSION THREAD

**Discussion by iamgoodbytes** [rappasoft/laravel-livewire-tables#1245] [Q&A] [answered]
Currently I have used the basic `protected $model = Product::class;` to load up the products in my database into the datatable. I'd like to only show my own records (that have a `user_id` matching the logged in user). Should I change the query that powers the source data?

https://rappasoft.com/docs/laravel-livewire-tables/v2/usage/the-query

**Comment by lrljoe** [comment]
You have three realistic options

First one (low impact - would recommend)
use the builder function to restrict the models returned.  Per the document you linked

Second one (potential impact)
Establish a global scope for the model, so no matter where the model is queried, it will only return "their" models.

Third one (re-engineer - more painful)
Adjust the model and queried items to return the items that "belong" to the user by querying the relevant relationship on the user

**Comment by iamgoodbytes** [comment]
@lrljoe thanks for your reply!

I'm trying the first, which seems easy to do but triggers the following error, might be just me not doing things the right way but can't seem to find what's wrong. Just tried to add the builder function like shown in the docs.

<img width="645" alt="Screenshot 2023-06-09 at 13 00 12" src="https://github.com/rappasoft/laravel-livewire-tables/assets/3899360/62e90176-4191-448a-9ef3-5addbdce43f0">

For the second or third option, are you referring to this technique? https://www.tutsmake.com/laravel-global-scope-example/

**Reply by lrljoe** [reply, marked as answer]
Make sure to include builder in the top of your component.
```php
use Illuminate\Database\Eloquent\Builder;
```

Don't typehint it or try to add any variables into the class instantiation

It should resemble
```php
    public function builder(): Builder
    {
        return User::query()
            // Your criteria here
    }
```

**Comment by iamgoodbytes** [comment]
Figured it out by checking the source of the original class

```
public function builder(): Builder
{
    return Product::query()
    ->where('user_id', auth()->id()); // Filter records by user_id
} 

Thanks for pointing me in the right direction!
```

**Reply by lrljoe** [reply]
Looks right to me!

Make sure you've added an index for user_id in your database to keep the query speedy!

For scopes, it may be worth looking at a local scope in a trait, and adding it to the various models, so that if you ever change the model from user to another authenticable, you make the change in one place.

I'd typically recommend that approach over repeating the same "where userid is the logged in user" throughout your code.
